src/transform/src/normalize_lets.rs

```diff
@@ -86,281 +88,576 @@ impl NormalizeLets {
+        let mut bindings = let_motion::harvest_non_recursive(relation);
+        if let MirRelationExpr::LetRec {
+            ids,
+            values,
+            body: _,
+        } = relation
+        {
+            bindings.extend(ids.drain(..).zip(values.drain(..)));
+            let (new_ids, new_values) = bindings.into_iter().unzip();
+            *ids = new_ids;
+            *values = new_values;
+        } else {
+            for (id, value) in bindings.into_iter().rev() {
+                *relation = MirRelationExpr::Let {
+                    id,
+                    value: Box::new(value),
+                    body: Box::new(relation.take_dangerous()),
+                };
+            }
         }
 
         Ok(())
     }
 }
 
-/// Re-assign an identifier to each `Let`.
-pub fn renumber_bindings(
-    relation: &mut MirRelationExpr,
-    id_gen: &mut IdGen,
-) -> Result<(), crate::TransformError> {
-    let mut renaming = BTreeMap::new();
-    let recursion_guard = RecursionGuard::with_limit(RECURSION_LIMIT);
-    support::renumber_bindings_helper(&recursion_guard, relation, &mut renaming, id_gen)
-}
+pub use renumbering::renumber_bindings;
 
 // Support methods that are unlikely to be useful to other modules.
 mod support {
 
     use std::collections::BTreeMap;
 
     use mz_expr::{Id, LocalId, MirRelationExpr};
-    use mz_ore::id_gen::IdGen;
-    use mz_ore::stack::{CheckedRecursion, RecursionGuard};
 
-    // This is pretty simple, but it is used in two locations and seemed clearer named.
-    /// Populates `counts` with the number of uses of each local identifier.
+    /// Logic mapped across each use of a `LocalId`.
+    pub(super) fn for_local_id<F>(expr: &MirRelationExpr, mut logic: F)
+    where
+        F: FnMut(LocalId),
+    {
+        expr.visit_pre(|expr| {
+            if let MirRelationExpr::Get {
+                id: Id::Local(i), ..
+            } = expr
+            {
+                logic(*i);
+            }
+        });
+    }
+
+    /// Populates `counts` with the number of uses of each local identifier in `expr`.
     pub(super) fn count_local_id_uses(
         expr: &MirRelationExpr,
         counts: &mut std::collections::BTreeMap<LocalId, usize>,
     ) {
-        expr.visit_pre(|expr| {
+        for_local_id(expr, |i| *counts.entry(i).or_insert(0) += 1)
+    }
+
+    /// Visit `LetRec` stages and determine and update type information for `Get` nodes.
+    ///
+    /// This method errors if the scalar type information has changed (number of columns, or types).
+    /// It only refreshes the nullability and unique key information. As this information can regress,
+    /// we do not error if the type weakens, even though that may be something we want to look into.
+    pub(super) fn refresh_types(expr: &mut MirRelationExpr) -> Result<(), crate::TransformError> {
+        // TODO: consider a different visitation order, or iterating, to propagate types further.
```

Comment:
Ah, .. I think I missed this point. Definitely seems worth fixing. The goal was to do at least what we'd do on a stack of `Let` nodes, but I think you are right.

Reply:
This should now be addressed. I'm not certain how to test it though (how to set up expressions that need the cascading refresh because they were wrong).